Support adding users to T-SQL db_owner database role

thephantomthief opened the pull request:
### Description

This commit adds support to alter the db_owner role and add/drop users from it.

Whenever a new user, say u1, is added to db_owner role, we will internally create a new role u1_obj which will be owner of all the objects of u1. By doing so, we are able to work around the cyclic dependency of role membership in postgres. If u1 creates new objects, using the pre-existing post object creation hook, we will reassign ownership of u1's objects to u1_obj.

Similarly, if u1 is dropped from db_owner role, we will drop the u1_obj role and reassign ownership u1's objects from u1_obj role back to u1.

Task: BABEL-4899
Signed-off-by: Sharu Goel <goelshar@amazon.com>

### Test Scenarios Covered ###
* **Use case based -** Yes


* **Boundary conditions -** N/A


* **Arbitrary inputs -** N/A


* **Negative test cases -** Yes


* **Minor version upgrade tests -** Yes


* **Major version upgrade tests -** Yes


* **Performance tests -** N/A


* **Tooling impact -** N/A


* **Client tests -** N/A



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

thephantomthief committed:
Support adding users to db_owner role

This commit adds support to alter the db_owner role and add/drop users
from it.

Whenever a new user, say u1, is added to db_owner role, we will
internally create a new role u1_obj which will be owner of all the
objects of u1. By doing so, we are able to work around the cyclic
dependency of role membership in postgres. If u1 creates new objects,
using the pre-existing post object creation hook, we will reassign
ownership of u1's objects to u1_obj.

Similarly, if u1 is dropped from db_owner role, we will drop the u1_obj
role and reassign ownership u1's objects from u1_obj role back to u1.

thephantomthief committed:
Block adding of db roles to db_owner

thephantomthief committed:
Fix test

thephantomthief committed:
Allow db_owner member to drop database

thephantomthief committed:
Fix ALTER/DROP USER and add more tests

thephantomthief committed:
fixes

thephantomthief committed:
fixes

thephantomthief committed:
add variable to track if we are inside create builtin dbs

thephantomthief committed:
Test in user database

thephantomthief committed:
Revert "Test in user database"

This reverts commit 192e63e50db81f8c26648c296064da1e71057fa0.

thephantomthief committed:
refactor

thephantomthief committed:
Test and build fix

thephantomthief committed:
Fix tests

thephantomthief committed:
Revert "Revert "Test in user database""

This reverts commit 1b639f7a6b15d6305ca783589c9f0517992fbe95.

thephantomthief committed:
Add upgrade tests

thephantomthief committed:
fix dump/restore tests

thephantomthief committed:
Remove redundant tests

thephantomthief committed:
test fixes

thephantomthief committed:
update schedule

anju15bharti committed:
Support Securityadmin fixed server role (#67)

Added support for new fixed server role securityadmin. This role has following privileges with it:

Members of the securityadmin fixed server role can manage logins and their properties.
They can GRANT, DENY, and REVOKE server-level permissions.
securityadmin can also GRANT, DENY, and REVOKE database-level permissions if they have access to a database.

Original PR with all the comments babelfish-for-postgresql#2907

Issue resolved: BABEL-5040
Signed-off-by: ANJU BHARTI <abanju@amazon.com>

anju15bharti committed:
Support Securityadmin fixed server role (#67)

Added support for new fixed server role securityadmin. This role has following privileges with it:

Members of the securityadmin fixed server role can manage logins and their properties.
They can GRANT, DENY, and REVOKE server-level permissions.
securityadmin can also GRANT, DENY, and REVOKE database-level permissions if they have access to a database.

Original PR with all the comments babelfish-for-postgresql#2907

Issue resolved: BABEL-5040
Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 committed:
Support fixed database role db_accessadmin (#66)

Description
For review comments see babelfish-for-postgresql#2970

Support fixed database role db_accessadmin in babelfish.
Member of db_accessadmin can do the following

CREATE SCHEMA
CREATE/DROP USER
ALTER USER
They always have connect permission
Engine PR: babelfish-for-postgresql/postgresql_modified_for_babelfish#447
Extension PR: #66
Issues Resolved
[BABEL-5136]

tanscorpio7 committed:
Support fixed database role db_accessadmin (#66)

Description
For review comments see babelfish-for-postgresql#2970

Support fixed database role db_accessadmin in babelfish.
Member of db_accessadmin can do the following

CREATE SCHEMA
CREATE/DROP USER
ALTER USER
They always have connect permission
Engine PR: babelfish-for-postgresql/postgresql_modified_for_babelfish#447
Extension PR: #66
Issues Resolved
[BABEL-5136]

tanscorpio7 committed:
Support fixed database role db_accessadmin (#66)

Description
For review comments see babelfish-for-postgresql#2970

Support fixed database role db_accessadmin in babelfish.
Member of db_accessadmin can do the following

CREATE SCHEMA
CREATE/DROP USER
ALTER USER
They always have connect permission
Engine PR: babelfish-for-postgresql/postgresql_modified_for_babelfish#447
Extension PR: #66
Issues Resolved
[BABEL-5136]

shalinilohia50 committed:
Support db_datareader/db_datawriter (#68)

Implement database roles `db_datareader` and `db_datawriter`

* Members of the db_datareader fixed database role can read all data from all user tables and views. User objects can exist in any schema except sys and INFORMATION_SCHEMA.
* Members of the db_datawriter fixed database role can add, delete, or change data in all user tables. In most use cases, this role is combined with db_datareader membership to allow reading the data that is to be modified.

Engine PR: https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/98

Task: [BABEL-3883](https://jira.rds.a2z.com/browse/BABEL-3883)
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

shalinilohia50 committed:
Support db_datareader/db_datawriter (#68)

Implement database roles `db_datareader` and `db_datawriter`

* Members of the db_datareader fixed database role can read all data from all user tables and views. User objects can exist in any schema except sys and INFORMATION_SCHEMA.
* Members of the db_datawriter fixed database role can add, delete, or change data in all user tables. In most use cases, this role is combined with db_datareader membership to allow reading the data that is to be modified.

Engine PR: https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/98

Task: [BABEL-3883](https://jira.rds.a2z.com/browse/BABEL-3883)
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

shalinilohia50 committed:
Support db_datareader/db_datawriter (#68)

Implement database roles `db_datareader` and `db_datawriter`

* Members of the db_datareader fixed database role can read all data from all user tables and views. User objects can exist in any schema except sys and INFORMATION_SCHEMA.
* Members of the db_datawriter fixed database role can add, delete, or change data in all user tables. In most use cases, this role is combined with db_datareader membership to allow reading the data that is to be modified.

Engine PR: https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/98

Task: [BABEL-3883](https://jira.rds.a2z.com/browse/BABEL-3883)
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

Commit:
Support for fixed server role dbcreator (#65)

This role has following privileges with it:
Members of the dbcreator fixed server role can
1. Create database
2. Alter database (if has access to that database, i.e either owner of the database or has mapped user in it)
3. Drop any database.

Engine PR : https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/93

Issues Resolved
[BABEL-5115]

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Support for fixed server role dbcreator (#65)

This role has following privileges with it:
Members of the dbcreator fixed server role can
1. Create database
2. Alter database (if has access to that database, i.e either owner of the database or has mapped user in it)
3. Drop any database.

Engine PR : https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/93

Issues Resolved
[BABEL-5115]

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Support for fixed server role dbcreator (#65)

This role has following privileges with it:
Members of the dbcreator fixed server role can
1. Create database
2. Alter database (if has access to that database, i.e either owner of the database or has mapped user in it)
3. Drop any database.

Engine PR : https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/93

Issues Resolved
[BABEL-5115]

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

thephantomthief committed:
fix merge conflicts

thephantomthief committed:
update db_accessadmin test output

HarshLunagariya committed:
Support fixed database role db_securityadmin (#70)

This commit adds support for db_securityadmin fixed database roles. Members of db_securityadmin can perform following operations within a particular database:

create schema
create database roles
alter/drop user-defined database roles
manage permissions via grant/revoke statements
Engine Changes - amazon-aurora/postgresql_modified_for_babelfish#100

Issues Resolved
[BABEL-5135]

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Support fixed database role db_securityadmin (#70)

This commit adds support for db_securityadmin fixed database roles. Members of db_securityadmin can perform following operations within a particular database:

create schema
create database roles
alter/drop user-defined database roles
manage permissions via grant/revoke statements
Engine Changes - amazon-aurora/postgresql_modified_for_babelfish#100

Issues Resolved
[BABEL-5135]

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Support fixed database role db_securityadmin (#70)

This commit adds support for db_securityadmin fixed database roles. Members of db_securityadmin can perform following operations within a particular database:

create schema
create database roles
alter/drop user-defined database roles
manage permissions via grant/revoke statements
Engine Changes - amazon-aurora/postgresql_modified_for_babelfish#100

Issues Resolved
[BABEL-5135]

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

thephantomthief committed:
rebase over latest commits

thephantomthief committed:
Add missing upgrade changes

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
re run tests

Yvinayak07 committed:
Changing time difference for non daylight saving period in getdate test file. (#3074)

As daylight saving period ends minute difference between 'UTC' and 'US/Pacific' is changed from 420 to 480. So changed the time zone from 'US/Pacific' to 'Asia/Kolkata' making it independent from daylight saving period.

Task: No-JIRA

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

thephantomthief committed:
fix test

thephantomthief committed:
Merge branch 'server-database-roles-4-x' into sharu-dev-db_owner_amzn

thephantomthief committed:
Add comments and fix tests

tanscorpio7 commented:
Who will own schema if user runs
```
CREATE SCHEMA s1 AUTHORIZATION l1
```
and l1 is a member of db_owner. Will the internal role own the schema ?

tanscorpio7 commented:
The current grant on schema logic assumes only dbo and schema_owner can create objects in the schema.
Now any role which is a member of db_owner can create objects in any schema.

Question is will objects created by members of db_owner adhere to grants on schema ?
```
grant select on schema::s1 to l2
```
then as a member of db_owner create in that schema
```
create table s1.t (id int) <-- can l2 select on this ?
```

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	PlannedStmt 	*wrapper;
+
+	/* TSQL specific behavior */
+	if (sql_dialect != SQL_DIALECT_TSQL)
```
Should we also check if current connection is TDS CONN ?

HarshLunagariya replied:
+1

thephantomthief replied:
Added check for TDS connection

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	/* TSQL specific behavior */
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return;
+
```
We probably should have `CommandCounterIncrement();` here

thephantomthief replied:
Done

tanscorpio7 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -2965,6 +2965,9 @@ bbf_object_access_hook(ObjectAccessType access, Oid classId, Oid objectId, int s
 
 	if (access == OAT_POST_CREATE && classId == ProcedureRelationId)
 		revoke_func_permission_from_public(objectId);
+
+	if (access == OAT_POST_CREATE)
+		change_object_owner_if_db_owner();
```
Since we are doing this way, we could add test cases for all the pg_catalog which contains some metadata related to ownerId and the newly created object.

For example when during create schema we record dependency on owner and default acls for the schema which also includes the ownerId, so we can add tests for both.

Not expecting to see any issues here, but would be good to have test cases. Thanks

thephantomthief replied:
Added for more object types

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -440,6 +440,10 @@ SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_authid_login_ext', '')
 SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_authid_user_ext', '');
 SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_schema_permissions', '');
 
+CREATE OR REPLACE FUNCTION sys.bbf_is_role_member(member NAME, rolename NAME)
+RETURNS BOOLEAN
+AS 'babelfishpg_tsql', 'bbf_is_role_member' LANGUAGE C;
+
```
Is there any possibility of merging this with `bbf_is_member_of_role_nosuper` ?

thephantomthief committed:
Address review comments

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -818,19 +820,27 @@ drop_bbf_db(const char *dbname, bool missing_ok, bool force_drop)
 				 errmsg("Cannot drop database \"%s\" because it is currently in use", dbname)));
 
 	/* Set current user to session user for dropping permissions */
+	prev_current_user_id = GetSessionUserId();
```
Better naming would be prev_session_user_id

thephantomthief replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3069,7 +3069,15 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							stmt->options = list_concat(stmt->options,
 														user_options);
-							create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
+
+							/*
+							 * If the role is created internally as part of ALTER ROLE
+							 * db_owner ADD MEMBER ... statement, we should not add this to
+							 * our babelfish catalog. These roles are meant to be internal
+							 * and not be visible to customer from Babelfish endpoint.
+							 */
+							if (strcmp(queryString, "(ALTER ROLE ADD )") != 0)
+								create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
```
We are only creating _obj role when adding any user as member of db_owner , right? Then should we have some kind of metadata just to rely on it to decide -  for which user, have we already created _obj_role and for which user, we haven't?

thephantomthief replied:
As discussed offline, if member of db_owner is dropped from db_owner role or the member user is dropped from database, we will also drop the corresponding _obj role

tanscorpio7 replied:
MACRO for `(ALTER ROLE ADD )`

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3955,13 +3963,19 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
-
+							else
+								standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
+														queryEnv, dest, qc);
```
nit: Let's utilise - `call_prev_ProcessUtility()` here

thephantomthief replied:
Done

thephantomthief committed:
add tests and address comments

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1603,48 +1606,93 @@ alter_bbf_authid_user_ext(AlterRoleStmt *stmt)
+		Oid 	db_owner_id, old_username_id;
+		char	*old_obj_rolname = NULL;
+		char	*new_obj_rolname = NULL;
+		bool	is_db_owner_member = false;
 
 		initStringInfo(&query);
 		appendStringInfo(&query, "ALTER ROLE dummy RENAME TO dummy; ");
 
+		db_owner_id = get_role_oid(get_db_owner_name(get_cur_db_name()), false);
+		old_username_id = get_role_oid(stmt->role->rolename, false);
+
+		/*
+		 * If user is member of db_owner role, we will also have to
+		 * rename the associated internal object owning role as well
+		 */
+		if (is_member_of_role(old_username_id, db_owner_id))
+		{
+			old_obj_rolname = get_obj_role(stmt->role->rolename);
+			new_obj_rolname = get_obj_role(physical_name);
+
+			appendStringInfo(&query, "ALTER ROLE dummy RENAME TO dummy; ");
+
+			is_db_owner_member = true;
+		}
+
 		parsetree_list = raw_parser(query.data, RAW_PARSE_DEFAULT);
 
-		if (list_length(parsetree_list) != 1)
+		if (!is_db_owner_member && list_length(parsetree_list) != 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("Expected 1 statement but got %d statements after parsing",
+							list_length(parsetree_list))));
+		else if (is_db_owner_member && list_length(parsetree_list) != 2)
 			ereport(ERROR,
 					(errcode(ERRCODE_SYNTAX_ERROR),
-					 errmsg("Expected 1 statement but get %d statements after parsing",
+					 errmsg("Expected 2 statement but got %d statements after parsing",
 							list_length(parsetree_list))));
```
Let's refactor this code to avoid duplication. Instead of 1/2 let's maintain a variable.

thephantomthief replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1838,13 +1887,21 @@ check_alter_role_stmt(GrantRoleStmt *stmt)
+	/*
+	 * Members of db_owner role can ALTER ANY ROLE
+	 */
+	if (has_privs_of_role(GetSessionUserId(), db_owner))
+		return;
```
Shouldn't it be CurrentUser?

thephantomthief replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+
+	initStringInfo(&rolname_obj);
+
+	appendStringInfoString(&rolname_obj, rolname);
+	appendStringInfoString(&rolname_obj, "_obj");
+
```
Let's make sure we add testcases for long username and database name.

thephantomthief replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+		Node		*n;
+		PlannedStmt	*wrapper;
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "DROP ROLE dummy; ");
```
Should we execute `DROP OWNED BY` statement here as well?

thephantomthief replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+/*
+ * The bbf_is_role_memeber function will check if a user (u1) is member of a role (r1):
+ * - If role is not a member of db_owner, we will check if u1 is member of r1 role
+ * - If role is a member of db_owner, we will check if u1 is member of r1_obj role
```
> If role is a member of db_owner, we will check if u1 is member of r1_obj role
We need to test this thoroughly in terms of metadata visibility as well.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	if (!user_exists_for_db(cur_db_name, rolname))
+		return;
+
+	if (!is_member_of_role(role_oid, get_role_oid(get_db_owner_name(cur_db_name), true)))
```
missing ok is true here, is it expected?

thephantomthief replied:
Made it false

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	}
+
+	systable_endscan(sscan);
+	table_close(rel, AccessShareLock);
+
+	/* Ownership of rest of the objects will be taken care of by REASSIGN OWNED */
+	appendStringInfoString(&query, "REASSIGN OWNED BY dummy TO dummy; ");
+	appendStringInfoString(&query, "DROP ROLE dummy; ");
```
REASSIGN OWNED BY doesn't cover schema object?

thephantomthief replied:
It does, but I noticed an inconsistency where sometimes we will first attempt to transfer ownership of say a procedure before its schema which was causing this operation to fail inconsistently. To workaround this, I am explicitly transferring ownership of schemas first and then running REASSIGN OWNED BY

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	appendStringInfoString(&query, "GRANT dummy TO dummy; ");
+
+	/* We will first transfer ownership of all the schemas owned by user */
+	rel = table_open(NamespaceRelationId, AccessShareLock);
```
RowExclusiveLock?

thephantomthief replied:
For read-only catalog access, AccessShareLock should suffice right?

shalinilohia50 replied:
Yes.

HarshLunagariya replied:
I was suggesting this because this row is being changed here -- We are changing the owner. Anyways, Engine API will anyway hold row ex. lock internally. so no worries to handle it here.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	appendStringInfoString(&query, "REVOKE dummy FROM dummy; ");
+
+	/* We will first transfer ownership of all the schemas owned by user */
+	rel = table_open(NamespaceRelationId, AccessShareLock);
```
RowExclusiveLock?

shalinilohia50 replied:
It should be `AccessShareLock` for read only.

thephantomthief committed:
Check for tds connection

Signed-off-by: Sharu Goel <goelshar@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5719,3 +5725,63 @@ handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId,
+			Oid 	db_owner = get_db_owner_oid(db_name, false);
+			Oid 	schema_db_id = get_dbid_from_physical_schema_name(NameStr(nsptup->nspname), false);
+
+			if (schema_db_id == get_cur_db_id() &&
+				has_privs_of_role(GetUserId(), db_owner))
+				ownerId = nsp_owner;
```
Let's skip this if current_user is dbo. For dbo we can fix it with BABEL-5361

thephantomthief replied:
Done

thephantomthief committed:
Address comments

thephantomthief committed:
Address comments and add tests for long db and user name

thephantomthief committed:
more fixes

shalinilohia50 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5719,3 +5725,65 @@ handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId,
+ * Objects are always owned by current user in postgres but in babelfish
+ * schema contained objects should be owned by the schema owner by default
+ * Use this hook to pick schema owner as object owner during object creation
+ * We currently only do this if current user is member of db_ddladmin
```
```
We currently only do this if current user is member of db_ddladmin
```
This should be for db_owner as well, right?

thephantomthief replied:
Yeah, actually I just copy pasted this from db_ddladmin to test if the same solution will work. My changes will be rebased over db_ddladmin commit so I will fix the comment when I'll get be resolving the conflicts.

shalinilohia50 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
 }
+
+char*
+get_obj_role(const char *rolname)
```
We create a role "rolname_obj" when a rolename is added as a member of db_owner. 
Are we making sure that a User cannot create this "_obj" role on its own?

thephantomthief replied:
We are not blocking it. There will be a limitation where if we already have say users u1_obj and u1, then end user won't be able to add u1 to db_owner role and we will throw an error. We had talked about this during design discussion and it was decided doing it like this is not really a big limitation

thephantomthief committed:
Add tests for more object types, DMLs, sp_rename, physical memberships

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Test fixes

Signed-off-by: Sharu Goel <goelshar@amazon.com>

tanscorpio7 committed:
Support fixed database role db_ddladmin (#71)

Support fixed database role db_ddladmin

Engine PR : amazon-aurora/postgresql_modified_for_babelfish#101

Task : BABEL-5116

Signed-off-by: Tanzeel Khan tzlkhan@amazon.com
Co-authored-by: Harsh Lunagaria lunharsh@amazon.com

tanscorpio7 committed:
Support fixed database role db_ddladmin (#71)

Support fixed database role db_ddladmin

Engine PR : amazon-aurora/postgresql_modified_for_babelfish#101

Task : BABEL-5116

Signed-off-by: Tanzeel Khan tzlkhan@amazon.com
Co-authored-by: Harsh Lunagaria lunharsh@amazon.com

tanscorpio7 committed:
Support fixed database role db_ddladmin (#71)

Support fixed database role db_ddladmin

Engine PR : amazon-aurora/postgresql_modified_for_babelfish#101

Task : BABEL-5116

Signed-off-by: Tanzeel Khan tzlkhan@amazon.com
Co-authored-by: Harsh Lunagaria lunharsh@amazon.com

thephantomthief committed:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into sharu-dev-db_owner_amzn

thephantomthief committed:
Fixes over db_ddladmin

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Merge branch 'server-database-roles-4-x' into sharu-dev-db_owner_amzn

thephantomthief committed:
Add order by and undo actions change

Signed-off-by: Sharu Goel <goelshar@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	 * Don't change object owner if current user is
+	 * dbo or we are creating system databases
+	 */
+	if (role_oid == dbo_id || dbo_id == InvalidOid || is_create_bbf_builtin_dbs)
```
Why is this flag needed, wouldn't we be in non tds connection when creating built in dbs ?

thephantomthief replied:
Yeah TDS_CONN check should suffice

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+
+		pfree(query.data);
+
+		if (obj_rolname)
```
can this be null in some cases ?

thephantomthief replied:
No. Removed the if()

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
+	Oid	db_owner_id;
+	char	*obj_rolname = NULL;
+
+	db_owner_id = get_role_oid(get_db_owner_name(get_cur_db_name()), false);
```
can we use `get_db_owner_oid`

thephantomthief replied:
Yes. Done

tanscorpio7 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5807,9 +5811,11 @@ pltsql_get_object_owner(Oid namespaceId, Oid ownerId)
-				!has_privs_of_role(GetUserId(), db_owner) &&
+				(has_privs_of_role(GetUserId(), db_owner) ||
 				has_privs_of_role(GetUserId(), db_ddladmin))
+				&& user_oid != dbo_oid)
```
This should be `ownerId` instead of user_oid. Some places will have ownerId and `GetUserId()` different even in engine and lets move this check before has_privs_of_role for efficiency,

thephantomthief replied:
Done

tanscorpio7 commented on contrib/babelfishpg_tsql/src/multidb.c:
```diff
@@ -268,11 +253,27 @@ rewrite_object_refs(Node *stmt)
+				physical_principal_name = get_physical_user_name(db_name, principal_name, false, true);
+
+				/* Forbidden ALTER ROLE db_owner ADD/DROP MEMBER if MEMBER is a T-SQL database role */
+				if ((strcmp(role_name, "db_owner") == 0) && get_db_principal_kind(get_role_oid(physical_principal_name, false), db_name) == BBF_ROLE)
```
MACRO for "db_owner" is defined.

thephantomthief replied:
Replaced with DB_OWNER

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3742,6 +3751,13 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							alter_owner = true;
 						}
+
+						if (has_privs_of_role(owner_oid, get_db_owner_oid(db_name, false)) && owner_oid != get_role_oid(get_dbo_role_name(db_name), false))
```
Let us have a helper for getting the dbo role oid given the db name. This will help reduce changes in multiple places.

thephantomthief replied:
Done

HarshLunagariya reviewed (commented):
Code changes looks good. There are failures in some of workflows.

thephantomthief committed:
Address review comments

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Merge branch 'server-database-roles-4-x' into sharu-dev-db_owner_amzn

thephantomthief committed:
Fix bad merge

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Add missing arg

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Fix db_accessadmin merge conflict

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into sharu-dev-db_owner_amzn

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -342,16 +342,18 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 	{
 		char	   *user_name = (char *) lfirst(elem);
-		char 	   *original_user_name = get_authid_user_ext_original_name(user_name, dbname);
+		char 	   *original_user_name = get_authid_user_ext_original_name(user_name, dbname, true);
```
When do we get original_user_name NULL?

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3069,7 +3069,15 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+							 * our babelfish catalog. These roles are meant to be internal
+							 * and not be visible to customer from Babelfish endpoint.
+							 */
+							if (strcmp(queryString, INTERNAL_ALTER_ROLE) != 0)
+								create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
```
Let's track the effort somewhere to add this constraint that "if any user is member of db_owner then it should have _obj user created" in metadata consistency check.

HarshLunagariya commented on test/JDBC/expected/db_owner-vu-verify.out:
```diff
@@ -0,0 +1,2009 @@
+~~END~~
+
+
+SELECT r.rolname AS parent_role
+FROM pg_auth_members m
+JOIN pg_roles r ON (m.roleid = r.oid)
+JOIN pg_roles mr ON (m.member = mr.oid)
+WHERE mr.rolname = 'dbowner__main_db_dbowner__u1'
+ORDER BY r.rolname;
+go
+~~START~~
+name
+dbowner__main_db_db_owner
+dbowner__main_db_dbo
+dbowner__main_db_dbowner__u1_obj
```
Have we tested `is_member('dbowner__main_db_dbowner__u1', 'dbo')` from ideal behaviour P.O.V. ?

HarshLunagariya replied:
+ just double checking if we have existing case covering for sp_helpuser, sp_helprolemember,... etc procedures.
+ We need to test extended properties as well.
